doc/man1/roarmonhttp.1
.\" roarmonhttp.1:

.TH "roarmonhttp" "1" "January 2010" "RoarAudio" "System User's Manual: roarmonhttp"

.SH NAME

roarmonhttp \- CGI and inetd based streaming server emulation for RoarAudio

.SH SYNOPSIS

roarmonhttp [\-\-inetd]

.SH DESCRIPTION

This program emulates a streaming server using RoarAudio.
It can be run from via a webserver as CGI or via inetd in inetd mode.
IN CGI mode it is limited to client (listener) HTTP connections.
In inetd mode it can handle source clients and gopher, too.
There is no need for a specal port for gopher. The protocol
is detected per client at runtime.

.SH "OPTIONS"

.TP
\fB--inetd\fR
Enables inetd mode.

.TP
\fB--server SERVER\fR
Set Server to connect to. This may be usfull in inetd mode.

.TP
\fB--rate\fR, \fB--bits\fR, \fB--channels\fR, \fB--codec\fR
Sets the default rate, bits, channels or codec.
Those option taking a argument for the given option. Codec names may be used.

.TP
\fB--rel-id SID\fR
Set the stream ID for the stream this stream is relative to.

.TP
\fB--help\fR
Show a brief help.

.SH "EXAMPLES"
Here is an example of a roarmonhttp listening on port 8000 using inetd:
 8000 stream tcp nowait roard /usr/bin/roarmonhttp roarmonhttp \-\-inetd

.SH "SEE ALSO"
\fBinetd\fR(8),
\fBinetd.conf\fR(5),
\fBroartips\fR(7),
\fBlibroar\fR(7),
\fBRoarAudio\fR(7).

.SH "HISTORY"

For history information see \fBRoarAudio\fR(7).

.\" ll


ChangeLog
v. 0.3 - ?
	Prereleases: 0: Mon Aug 16 2010 29:58 CEST
	* Fixed locking bug with libroaross and Audacity
	* Added working resampler code thanks to maister.
	  (Closes: DEB#592017)
	* Changed way to find default driver.
	* Updated Debian init script
	* Added VS API
	* Fixed strange bug in RSound emulation killing streams
	  after long time.
	* Removed libao and audacious plugins (they are in upstream now)
	* Added basic bash completion script

v. 0.3beta7 - Fri Jul 23 2010 15:11 CEST
	Prereleases: 0: Mon Jul 19 2010 12:58 CEST;
	             1: Mon Jul 19 2010 20:37 CEST;
	             2: Wed Jul 21 2010 11:07 CEST
	* Added support for audio profiles
	* Updated libroarrsound
	* Added RPlay protocol to roard
	* Updates of copyright and license headers
	* Added ckport file
	* Updated some tools to get away from old, legacy
	  and security relevant function calls
	* Fixed bug with incorrect number of arguments to roard
	* Added some more ROAR_INFO()s to roard
	* Added tool roarclientpass
	* Corrected Debian init script (pr1)
	* Fixed problems on Big Endian systems with libdnet (pr2)

v. 0.3beta6 - Thu Jun 24 2010 20:47 CEST
	Prereleases: 0: Fri Jun 11 2010 10:47 CEST;
	             1: Sat Jun 12 2010 11:07 CEST;
	             2: Wed Jun 16 2010 12:52 CEST
	* Added something like select() for VIOs
	* Fixed a lot warnings
	* disabled cmd codecfilter by default
	* Fixed error with roarinterconnect on non-OSS systems
	* Fixed some problems with minimal builds
	* Some Copyright fixes
	* much better VIO support for connections
	* much better OSS emulation
	* Updated win32 port
	* Updated RSound support
	* fixed pos calc in case of PCM->PCM converting (pr1)
	* fixed libroarpulse (pr2)

v. 0.3beta5 - Sun May 02 2010 12:41 CEST
	* added tool roardtmf
	* Support to store server address in X11's root window propertys
	* some small fixes
	* wrote a lot more libroarpulse code, some async API clients work now
	* added compbins for esddsp, artsdsp, audiooss, padsp
	* splited libroarpulse and libroarpulse-simple
	* upgraded some code to use VIO
	* more updates for Debian
	* fixed init script for Debian and Debian like systems
	* fixed some mixer stuff, a lot things work better now
	* added support for notify beeps
	* updated drivers
	* updated ROAR_{ERR,WARN,INFO,DBG}()
	* added syslog support
	* updated test suit
	* done a lot work on libroaross
	* fixed segfault in ALSA plugin
	* Added support for OSS to roarinterconnect(1)
	* Added server side RSound Emulation
	* Added libroarrsound, client side RSound Emulation
	* Added ALSA driver
	* Added RSound driver
	* Added PortAudio driver

v. 0.3beta4 - Sat Feb 06 2010 26:02 CET
	* Applyed patches for Debian
	* updated roarify to load libs of correct version
	* updated list of supported versions (complibs)
	* added header with common units
	* added support for PulseAudio Simple protocol
	* added support to libroaross to emulate DMX4Linux (write only)
	* added DSTR driver to roard
	* updated librarys' soname
	* do not link OpenSSL by default (licence incompatible?)
	* support for dynamic lib loading, plugin interface
	* support to for registering DSTR types
	* moved WinSock things out of main vio.[ch]
	* support auto-tool --prefix*=*

v. 0.3beta3 - Fri Jan 22 2010 22:17 CET
	* added some aRts/NAS/YIFF/PulseAudio tools to compbins
	* added support for comp libs to roar-config
	* added support for uni-directional streams in roarinterconnect
	* write much better support for config, including codec config
	* reduced the TCP overhead produced by speex
	* added support for speex preprocessor
	* get names of complibs from table, do not guess them
	* a lot small fixes
	* moved most of the mixer stuff out of roard into libroardsp
	* added flags immutable and enhance
	* test for signal.h
	* Corrected (DSTR) HTTP Protocol support so servers without headerflush
	  should now work, too.
	* added protocol support for icy
	* Added version numbers to lib names
	* updated manpages
	* wrote OSS emulation
	* fixed test scripts

v. 0.3beta2 - Mon Oct 05 2009 19:48 CEST
	* Added prethru flag
	* Added support to read info about client's network node
	* Ported libroaresd, libroarartsc
	* Fixed some problems with EsounD emulation
	* added support for compatibility binarys
	* added poll() support to libroarsndio
	* reject streams of unsupported codec

v. 0.3beta1 - Sat Sep 26 2009 25:50 CEST
	* incremented nummber of streams per client fron 4 to 8
	* RIFF WAVE codecfilter uses virtual streams
	* another set of changed names of parameters in roarctl
	* Interface to read byteorder of a client from roard
	* implemented a 'state of stream' var
	* cleanup, small fixes
	* added VIO interface to transcoder interface
	* ask codecfilter of parent in case of new child stream
	* added 'subsystem' complex: multi subsystem data on single streams
	* added basic /Radio Data and Trasmitter Control System/.
	* better support for stereo upmixing
	* make audio group a optional thing
	* better support for win32
	* wrote WMM driver
	* better multi protocol support
	* Support for server side EsounD emulation on win32

v. 0.3beta0 - Sun Sep 06 2009 22:42 CEST
	* added new stream directions for MIDI and light control
	* cleanup of client stream struct
	* added basic light control subsystem
	* added DMX512 driver
	* wrote basic MIDI subsystem
	* added flag 'mute'
	* added flag 'mmap'
	* added lib libroarmidi and libroarlight
	* added OpenSLP support
	* rewote fillmixbuer2 code
	* added a roar_conv2() with a more clear API
	* added support to write pidfile
	* added support for encoding 32 bit Vorbis streams
	* added transcoding features in libroardsp
	* added tool 'roarphone'
	* added tool 'roarinterconnect'
	* added simple raw data subsystem
	* added working support for THRU streams
	* changed names of parameters in roarctl
	* added lib libroareio - Extended IO for RoarAudio
	* Update order of default drivers, now it is: oss, ao, esd, null
	* cleanup of old sysio driver interface

roarclients/roarmonhttp.c
//roarmonhttp.c:

/*
 *      Copyright (C) Philipp 'ph3-der-loewe' Schafft - 2008-2010
 *
 *  This file is part of roarclients a part of RoarAudio,
 *  a cross-platform sound system for both, home and professional use.
 *  See README for details.
 *
 *  This file is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License version 3
 *  as published by the Free Software Foundation.
 *
 *  RoarAudio is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this software; see the file COPYING.  If not, write to
 *  the Free Software Foundation, 51 Franklin Street, Fifth Floor,
 *  Boston, MA 02110-1301, USA.
 *
 */

#include <roaraudio.h>

#if defined(ROAR_HAVE_SETENV) || defined(ROAR_HAVE_PUTENV)
#define _CAN_SET_ENV
#endif

#define BUFSIZE 1024

void usage (void) {
 printf("roarmonhttp [OPTIONS]...\n");

 printf("\nOptions:\n\n");

 printf("  --server    SERVER    - Set server hostname\n"
        "  --rate  -R  RATE      - Set sample rate\n"
        "  --bits  -B  BITS      - Set bits per sample\n"
        "  --chans -C  CHANNELS  - Set number of channels\n"
        "  --codec     CODEC     - Set the codec\n"
        "  --rel-id ID           - Set ID of relative stream\n"
        "  --inetd               - Start in inetd mode (STDIN and STDOUT connected to socket)\n"
        "  --help                - Show this help\n"
       );

}

void print_header (int codec, int rate, int channels) {
 char * mime = "application/octet-stream";

 switch (codec) {
  case ROAR_CODEC_OGG:
  case ROAR_CODEC_OGG_GENERAL:
  case ROAR_CODEC_OGG_VORBIS:
    mime = "application/ogg";
   break;
  case ROAR_CODEC_OGG_SPEEX:
    mime = "audio/ogg; codecs=speex";
   break;
  case ROAR_CODEC_OGG_FLAC:
    mime = "audio/ogg; codecs=flac";
   break;
  case ROAR_CODEC_OGG_CELT:
    mime = "audio/ogg; codecs=celt";
   break;
  case ROAR_CODEC_FLAC:
    mime = "audio/x-flac";
   break;
  case ROAR_CODEC_RIFF_WAVE:
    mime = "audio/x-wav";
   break;
  case ROAR_CODEC_AU:
    mime = "audio/basic";
   break;
  case ROAR_CODEC_AIFF:
    mime = "audio/aiff";
   break;
 }

 printf("Content-type: %s\r\n", mime);
 printf("ice-audio-info: ice-samplerate=%i;ice-channels=%i\r\n", rate, channels);
 printf("icy-pub:0\r\n");
 printf("Server: RoarAudio (roarmonhttp $Revision: 1.23 $)\r\n");
 printf("\r\n");

 fflush(stdout);
}

int stream (int dest, int src) {
 struct roar_buffer *ring = NULL, *cur;
 ssize_t len;
 size_t  todo;
 fd_set fsi[1], fso[1];
 struct timeval tv;
 int alive = 1;
 int maxfh = (dest > src ? dest : src) + 1;
 void * data;

 roar_socket_nonblock(src,  ROAR_SOCKET_NONBLOCK);
 roar_socket_nonblock(dest, ROAR_SOCKET_NONBLOCK);

 while (alive) {
  FD_ZERO(fsi);
  FD_ZERO(fso);
  FD_SET(src, fsi);
  if ( ring != NULL ) {
   FD_SET(dest, fso);
  }

  tv.tv_sec  = 0;
  tv.tv_usec = 100000; // 100ms

  if (select(maxfh, fsi, fso, NULL, &tv) > 0) {
   if ( FD_ISSET(src, fsi) ) { // we can read!
    if ( roar_buffer_new(&cur, BUFSIZE) == -1 )
     return -1;

    if ( roar_buffer_get_data(cur, &data) == -1 )
     return -1;

    len = read(src, data, BUFSIZE);

    switch (len) {
     case  0:
     case -1:
       roar_buffer_free(cur);

       if ( ring != NULL )
        roar_buffer_free(ring);

       return -1;
      break;
    }

    if ( roar_buffer_set_len(cur, len) == -1 )
     return -1;

    if ( ring == NULL ) {
     ring = cur;
    } else {
     roar_buffer_add(ring, cur);
    }
   } else if ( FD_ISSET(dest, fso) && ring != NULL ) { // we can write!
    if ( roar_buffer_get_data(ring, &data) == -1 )
     return -1;

    if ( roar_buffer_get_len(ring, &todo) == -1 )
     return -1;

    len = write(dest, data, todo);

    if ( len < 1 ) {
     if ( errno != EAGAIN ) {
      roar_buffer_free(ring);
      return -1;
     }
    }

    if ( (ssize_t)todo == len ) { // we wrote all of the pkg
     if ( roar_buffer_next(&ring) == -1 )
      return -1;
    } else {
     if ( roar_buffer_set_offset(ring, len) == -1 )
      return -1;
    }

   }
  }
 }

 return 0;
}

#ifdef _CAN_SET_ENV
int parse_http (int * gopher) {
 struct roar_keyval kv;
 char buf[1024];
 char * qs = buf, *str;
 ssize_t len;
 int dir = ROAR_DIR_MONITOR;

 if ( (len = read(ROAR_STDIN, buf, 1023)) == -1 )
  return -1;

 buf[len] = 0;

 if ( strncmp(buf, "GET /", 5) ) {
  if ( strncmp(buf, "SOURCE /", 8) ) {
   if ( buf[0] != '/' ) {
    return -1;
   } else {
    *gopher = 1;
   }
  } else {
   dir = ROAR_DIR_PLAY;
   qs += 3; 
  }
 }

 if ( !*gopher ) {
  qs += 5;

  if ( (str = strstr(qs, " ")) == NULL )
   return -1;

  *str = 0;
 } else {
  if ( (str = strstr(qs, "\r")) != NULL )
   *str = 0;
  if ( (str = strstr(qs, "\n")) != NULL )
   *str = 0;
 }

 for (; *qs != '?'; qs++)
  if ( !*qs )
   break;

 if ( *qs == '?' )
  qs++;

 if ( !*gopher )
  printf("HTTP/1.0 200 OK\r\n");
// printf("QS: %s\r\n", qs);

 fflush(stdout);

 kv.key   = "QUERY_STRING";
 kv.value = qs;

 roar_env_set(&kv);

 return dir;
}
#endif

int main (int argc, char * argv[]) {
 int    rate     = 44100;
 int    bits     = 16;
 int    channels = 2;
 int    codec    = ROAR_CODEC_OGG_VORBIS;
 int    rel_id   = -1;
 int    sflags   = ROAR_FLAG_NONE;
// int    codec    = ROAR_CODEC_DEFAULT;
 char * server   = NULL;
 int    fh;
 int    i;
 char * c, * k, * v;
#ifdef ROAR_HAVE_STRTOK_R
 char * sp0 = NULL, * sp1 = NULL;
#endif
 int dir = ROAR_DIR_MONITOR;
 int gopher = 0;
 struct roar_connection    con;
 struct roar_stream        s;

#ifdef ROAR_HAVE_ALARM
 alarm(0); // reset alarm timers from httpd 
#endif

 for (i = 1; i < argc; i++) {
  k = argv[i];
  if ( !strcmp(k, "--inetd") ) {
#ifdef _CAN_SET_ENV
   if ( (dir = parse_http(&gopher)) == -1 )
    return 1;
#else
   return 1;
#endif
  } else if ( !strcmp(k, "--server") ) {
   roar_libroar_set_server(argv[++i]);
  } else if ( !strcmp(k, "--codec") ) {
   codec = roar_str2codec(argv[++i]);
  } else if ( !strcmp(k, "--rate") || !strcmp(k, "-r") || !strcmp(k, "-R") ) {
   rate = atoi(argv[++i]);
  } else if ( !strcmp(k, "--bits") || !strcmp(k, "-B") ) {
   bits = atoi(argv[++i]);
  } else if ( !strcmp(k, "--channels") || !strcmp(k, "--chans") || !strcmp(k, "-C") ) {
   channels = atoi(argv[++i]);
  } else if ( !strcmp(k, "--rel-id") ) {
   rel_id = atoi(argv[++i]);
  } else if ( !strcmp(k, "--help") && !strcmp(k, "-h") ) {
   usage();
   return 0;
  } else {
   ROAR_ERR("Unknown parameter: %s", k);
   usage();
   return 1;
  }
 }

 c = getenv("QUERY_STRING");
 if ( c == NULL )
  c = "";

#ifdef ROAR_HAVE_STRTOK_R
 c = strtok_r(c, "&", &sp0);
#else
 c = strtok(c, "&");
#endif

 while (c != NULL) {
#ifdef ROAR_HAVE_STRTOK_R
  k = strtok_r(c,    "=", &sp1);
  v = strtok_r(NULL, "=", &sp1);
#else
  k = c;
  if ( (v = strstr(c, "=")) != NULL ) {
   *v = 0;
   v++;
  }
#endif

  if ( !strcmp(k, "codec") ) {
   if ( (codec = roar_str2codec(v)) == -1 )
    return 1;
  } else if ( !strcmp(k, "channels") ) {
   channels = atoi(v);
  } else if ( !strcmp(k, "rate") ) {
   rate = atoi(v);
  } else if ( !strcmp(k, "bits") ) {
   bits = atoi(v);
  } else if ( !strcmp(k, "rel-id") || !strcmp(k, "relid") ) {
   rel_id = atoi(v);
  } else if ( !strcmp(k, "set-flag") ) {
   if ( !strcmp(v, "meta") ) {
    sflags |= ROAR_FLAG_META;
   } else if ( !strcmp(v, "cleanmeta") ) {
    sflags |= ROAR_FLAG_CLEANMETA;
   } else if ( !strcmp(v, "prethru") ) {
    sflags |= ROAR_FLAG_PRETHRU;
   } else {
    return 1;
   }
  } else if ( !strcmp(k, "dir") ) {
   if ( (dir = roar_str2dir(v)) == -1 )
    return 1;
  } else {
   return 1;
  }

#ifdef ROAR_HAVE_STRTOK_R
  c = strtok_r(NULL, "&", &sp0);
#else
  c = strtok(NULL, "&");
#endif
 }

 if ( roar_simple_connect(&con, server, "roarmonhttp") == -1 ) {
  return 10;
 }

 if ( roar_stream_new(&s, rate, channels, bits, codec) == -1 ) {
  roar_disconnect(&con);
  return 20;
 }

 if ( rel_id != -1 ) {
  if ( roar_stream_set_rel_id(&s, rel_id) ) {
   roar_disconnect(&con);
   return 21;
  }
 }

 if ( roar_stream_connect(&con, &s, dir) == -1 ) {
  roar_disconnect(&con);
  return 11;
 }

 if ( sflags != ROAR_FLAG_NONE ) {
  if ( roar_stream_set_flags(&con, &s, sflags, 0) == -1 ) {
   roar_disconnect(&con);
   return 14;
  }
 }

 if ( roar_stream_exec(&con, &s) == -1 ) {
  roar_disconnect(&con);
  return 12;
 }

 if ( (fh = roar_get_connection_fh(&con)) == -1 )
  return 1;

 if ( !gopher )
  print_header(codec, rate, channels);

/*
 while((i = read(fh, buf, BUFSIZE)))
  if (write(out, buf, i) != i)
   break;
*/

 switch (dir) {
  case ROAR_DIR_PLAY:
    stream(fh, ROAR_STDIN);
   break;
  case ROAR_DIR_MONITOR:
  case ROAR_DIR_THRU:
    stream(ROAR_STDOUT, fh);
   break;
 }

 roar_simple_close(fh);

 return 0;
}

//ll
